### PR TITLE
Place the new marker when a chat becomes visible

### PR DESCRIPTION
**Pull request by Gonals**
@tgolen, @marcaaron, since we were all working on this. 
  
### Details
I decided to put the code on `onVisibilityChange` in `ReportActionsView` because that way we keep all the "similar" logic in the same file and I think it is really easy to understand and follow what's going on.
BUT, it could totally go [here](https://github.com/Expensify/Expensify.cash/blob/main/src/libs/actions/Report.js#L519) in `updateReportWithNewAction` too and it works just as well and it is also a logical place for it to be so... 🤷. If anyone has a strong preference, I'm fine with either.

### Fixed Issues

https://github.com/Expensify/Expensify/issues/163167

### Tests/ QA Steps
In web:
- Open a chat that does NOT have the New marker in place and move to a different tab.
- Receive two messages in the chat (wait for the notifications)
- Go back to the `.cash` tab. The new marker should be over the oldest of the new messages.
- Open a chat that DOES have the new marker (or place it marking a comment as unread) and move to a different tab.
- Receive a message in the chat (wait for the notification)
- Go back to the `.cash` tab. The New marker should be in the same place it was before.

In web and all the other platforms:
- Open a chat that does NOT have the New marker in place and minimize the window/app.
- Receive two messages in the chat (wait for the notifications)
- Go back to the `.cash` tab. The new marker should be over the oldest of the new messages.
- Open a chat that DOES have the new marker (or place it marking a comment as unread) and minimize the window/app.
- Receive a message in the chat (wait for the notification)
- Go back to the `.cash` tab. The New marker should be in the same place it was before.

### Tested On

- [X] Web
- [ ] Mobile Web
- [X] Desktop
- [X] iOS
- [X] Android

### Screenshots
Taking screenshots of the marker won't mean much.

### CODE REVIEW

**Review by marcaaron — COMMENTED** [review]
I guess my one question is do we need to hook into the visibility event? Doesn't this mean we will only set the new marker when the **user becomes active**.

I think this works, but I might expect this to work like this:

1. User is not active
2. New comments appear 
3. We should reset the new marker because they are not active

That would probably require adding similar logic to the code that handles the new actions here (which I think is the "event" we are reacting to here and maybe not the user becoming active):

https://github.com/Expensify/Expensify.cash/blob/a22e73de0086c2694026d545e15a8862af3990d5/src/pages/home/report/ReportActionsView.js#L155-L182

But maybe I'm missing something about why we would prefer to do it this way.

**Comment by tgolen** [comment]
Heh, @marcaaron I think that's the same thing I suggested to Alberto when he originally asked me about this.

@Gonals maybe since both Marc and I had that same instinct, it might be a good reason to go with that solution of this one. Do you know how much more complex that solution would be? I do like that this PR is nice and simple, and I don't have too good of a sense over how complex the other solution would be.

File: src/pages/home/report/ReportActionsView.js [hunk 1]
```diff
@@ -197,6 +197,12 @@ class ReportActionsView extends React.Component {
      */
     onVisibilityChange() {
         if (Visibility.isVisible()) {
+            // When the report becomes visible, if the New marker is not already set and we have unread actions,
+            // set the marker above the first (oldest) unread sequence.
```
[the remark this answers is not in the document]
Comment (Gonals) [reply src/pages/home/report/ReportActionsView.js:201]: Added extra info!

**Comment by Gonals** [comment]
All comments addressed!

File: src/libs/actions/Report.js [hunk 2]
```diff
@@ -516,6 +526,13 @@ function updateReportWithNewAction(reportID, reportAction) {
     if (lodashGet(reportAction, 'actorEmail') === 'concierge@expensify.com') {
         return;
     }
+
+    // When a new message comes in, if the New marker is not already set (newMarkerSequenceNumber === 0) and, set the
+    // marker above the incoming message.
+    if (lodashGet(allReports, reportID).newMarkerSequenceNumber === 0 && updatedReportObject.unreadActionCount > 0) {
```
Comment (marcaaron) [inline src/libs/actions/Report.js:532]: It's very possible that this report does not yet exist. And if `lodashGet(allReports, reportID)` is `undefined` then we will try and fail to access `newMarkerSequenceNumber`.

In the case that it doesn't exist I think we would treat that as the same as having a `newMarkerSequenceNumber === 0` so we should do this instead...

```suggestion
    if (lodashGet(allReports, [reportID, 'newMarkerSequenceNumber'], 0) === 0 && updatedReportObject.unreadActionCount > 0) {
```

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Review by tgolen — DISMISSED** [review]
I like it with Marc's suggested change.

File: src/libs/actions/Report.js (hunk 2 above)
Comment (Gonals) [reply src/libs/actions/Report.js]: Ah! That makes sense. Good catch!

**Review by marcaaron — APPROVED** [review]

**Pull request merged**